component/loki/source/file/file.go

```diff
@@ -0,0 +1,277 @@
+	if err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+	positionsFile, err := positions.New(o.Logger, positions.Config{
+		SyncPeriod:        10 * time.Second,
+		PositionsFile:     filepath.Join(o.DataPath, "positions.yml"),
+		IgnoreInvalidYaml: false,
+		ReadOnly:          false,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Component{
+		opts:    o,
+		metrics: newMetrics(o.Registerer),
+
+		handler:   make(chan api.Entry),
+		receivers: args.ForwardTo,
+		posFile:   positionsFile,
+		readers:   make(map[string]Reader),
+	}
+
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+// TODO(@tpaschalis). Should we periodically re-check? What happens if a target
+// comes alive _after_ it's been passed to us and we never receive another
+// Update()? Or should it be a responsibility of the discovery component?
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		level.Info(c.opts.Logger).Log("msg", "loki.source.file component shutting down, stopping readers")
+		for _, r := range c.readers {
+			r.Stop()
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler:
+			for _, receiver := range c.receivers {
+				receiver <- entry
+			}
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.args = newArgs
+	c.receivers = newArgs.ForwardTo
+
+	oldPaths := make(map[string]struct{})
+
+	// Stop all readers and recreate them below. This avoids the issue we saw
+	// with stranded wrapped handlers staying behind until they were GC'ed and
+	// sending duplicate message to the global handler. It also makes sure that
+	// we update everything with the new labels. Simply zeroing out the
+	// c.readers map did not work correctly to shut down the wrapped handlers
+	// in time.
+	// TODO (@tpaschalis) We should be able to optimize this somehow and eg.
+	// cache readers for paths we already know about, and whose labels have not
+	// changed. Once we do that we should:
+	// a) Call to c.pruneStoppedReaders to give cached but errored readers a
+	// chance to restart.
+	// b) Stop tailing any files that were no longer in the new targets
+	// and conditionally remove their readers only by calling toStopTailing
+	// and c.stopTailingAndRemovePosition.
+	for p, r := range c.readers {
+		oldPaths[p] = struct{}{}
+		r.Stop()
+	}
+	c.readers = make(map[string]Reader)
+
+	if len(newArgs.Targets) == 0 {
+		level.Debug(c.opts.Logger).Log("msg", "no files targets were passed, nothing will be tailed")
+		return nil
+	}
+
+	for _, target := range newArgs.Targets {
+		path := target[pathLabel]
+		c.reportSize(path)
+
+		var labels = make(model.LabelSet)
+		for k, v := range target {
+			if strings.HasPrefix(k, "__") {
```

Comment:
nit: use `model.ReservedLabelPrefix` (from github.com/prometheus/common/model) here instead of `"__"`.